### ISSUE TITLE
deduplication

### ISSUE THREAD

**Issue by Amarahtbase** [lizardfs/lizardfs#558]
Hello,

I would like an information please.
I would like to know if it's possible to do deduplication with lizardfs?
Thank you
Regards,
A.A

**Comment by Blackpaw** [comment]
No, not possible I'm afraid.

You could use zfs for the chunkserver disks and enable zfs deduplication. Note that deduplication is very memory intensive on zfs.

Also with zfs you can enbale on the fly lz4 compression, depending on your usecase might be better results. Much less memory consumption and actually improves io. I get 30% space savings with my VM storage.

**Comment by cloudweavers** [comment]
With ZFS I would suggest enabling deduplication only if you *really* need it. It has substantial performance implications, uses an inordinate amount of memory that makes sense only for very large storage and a huge degree of copying between files. Our platform is zfs-based, and after a few months of testing we found out that it is much more cost-effective to simply add more disks.
Compression is much more effective- has very limited impact on CPU, and provides substantial advantages in space and IO (less data to read and write)

**Comment by Blackpaw** [comment]
For sure, it seems to be a very specialised and limited use case.

**Comment by Amarahtbase** [comment]
Thank you all for you time ;)

**Comment by Amarahtbase** [comment]
I already know for the deduplication with zfs and I am looking a new solution that I will not know.

**Comment by onlyjob** [comment]
Speaking about compression I'm wondering what's holding implementation of #366?
IMHO native compression in chunkservers would be a big win for _xfs_ and _ext4_ with only small effort required to do the work...

**Comment by psarna** [comment]
@onlyjob 
There are 3 ways of implementing compression we discussed and they all have disadvantages:
1. Compressing data on disk only - but then, one can use ready solutions, no need of implementing it again
2. During network transfer only - but then, transfer is limited to compression/decompression speed, which isn't that great (i.e. 300MBps compression speed for lzo)
3. Keeping chunks compressed all the time, on disk and during network transfer - but then, it's not trivial how to handle requesting certain offsets from the chunk (decompressing the whole chunk for a single block looks redundant), and protocol changes must occur - clients and chunkservers should be able to know if the chunk is compressed or not.

So, to sum up, we don't think it is small effort, so the implementation is delayed until we have enough resources to research, implement and test it properly.

**Comment by Amarahtbase** [comment]
@psarna  Thank you @psarna 
How do you implement the second way? Could you give me an example?

Regards

**Comment by psarna** [comment]
@Amarahtbase
The second way would include changing lizardfs-client and lizardfs-chunkserver code, so it's not exactly easy to give you details and examples. Right now chunkserver reads needed blocks to a buffer and sends them, the change would simply be to apply lzo/lz4 or any other compression algorithm on the buffer first. Then a client would apply matching decompression algorithm on a received buffer. That's it.

**Comment by cypromis** [comment]
We actually happily use compression in ZFS on the chunkservers. Works like a charm. Same with the nodeweaver guys.
If it comes to deduplication, the best results for dedup are usually achieved on the application level since applications tend to know their file structures best which makes application level dedup far simpler than filesystem or blocklevel dedup.

**Comment by 4Dolio** [comment]
Missed this topic when it was occuring but wanted to add that i have used zfs compression on select chunk servers which store archival data. It works well, and may even work well for more 'live' data which is highly compressable. Another compression and metedata reducing method i use it to store a .zfs 'lookback' style disk image on LFS with compression enables within the disk image, which also works great and can store millions of objects while consuming very little metadata.  The drawback is only one client can use the pseudo-disk at a given time...

**Comment by Zorlin** [comment]
@4Dolio I consistently enjoy hearing about your adventures with LFS and the cool things you do with it. Thanks for sharing :)
